### PR TITLE
feat: add support for getting column names from Read Buffer

### PR DESCRIPTION
**Pull request by e-dard**
This PR implements the physical operation `column_names` on the Read Buffer. This operation can be used to retrieve a set of columns names for a single table across one or more Chunks of data. `column_names` accepts an optional predicate, which if set will ensure only column names are returned where at least one value in the column is on a row that matches the predicate.

This functionality will provide the execution  for IOx's equivalent of InfluxDB's `SHOW TAG KEYS WHERE ...` schema exploration query and Flux's `schema.tagKeys()`.

Here is an example using a pseudo query language:

```
Table: cpu

region  | env  | counter  | time
--------------------------------
west    | prod |    22    |  100
east    | prod |  NULL    |  200
west    | dev  |  NULL    |  300


SHOW COLUMN_NAMES FROM "cpu";

names
-------
counter
env
region
time

SHOW COLUMN_NAMES FROM "cpu" WHERE env = 'dev';

names
------- 
             <-- counter has no non-null values for matching rows
env
region
time
``` 

The usual current rules apply to the predicate provided the read buffer - conjunctive binary expressions only at the moment supporting `=, !=, >, <, <=, >=` on any column.

#### Implementation Notes

I have implemented `column_names` in a similar way to `table_names`, which I think will be performant enough for a while.

The implementation has the following features:

 - it works across chunks if multiple chunks are provided;
 - it maintains a set of matching column names, which are pushed down to each table and row group;
 - if a row group does not contain any columns that are not present in the matched result set it skips it;
 - if a column in a row group already exists in the result set it skips it.

These features means that in the general case across large swathes of data, `column_names` will be very efficient in the general case.

### CODE REVIEW

**Comment by alamb** [comment]
In the example

```

SHOW COLUMN_NAMES FROM "cpu" WHERE env = 'dev';

names
------- 
             <-- counter has no non-null values for matching rows
env
region
time
```

Do the results have 3 or 4 rows? (aka is the empty line meant to signify a row with "NULL" or that the result does not contain any row for `counter`?

File: read_buffer/src/lib.rs [hunk 1]
```diff
@@ -470,20 +474,42 @@ impl Database {
...
+            .fold(BTreeSet::new(), |dst, chunk| {
+                // the dst buffer is pushed into each chunk's `column_names`
+                // implementation ensuring that we short-circuit any tables where
+                // we have already determined column names.
+                chunk.column_names(table_name, &predicate, dst)
+            }) // have a BTreeSet here, convert to an iterator of Some(&str)
+            .into_iter()
+            .map(Some);
+
+        str_iter_to_batch(COLUMN_NAMES_COLUMN_NAME, names).context(ArrowError)
```
Comment (alamb) [inline read_buffer/src/lib.rs:512]: Just as FYI (no changes needed) I have been considering the general case when there are predicates that can't be pushed down to the storage system for "metadata queries" such as table_names and `column_names`. In that case I need to have a plan that pushes as many predicates down as possible and then do some extra filtering above. 

TLDR is that I am thinking of switching these APIs to return the raw `BTreeSet<String>` instead of a RecordBatch -- I don't think it will have much impact on this code but I wanted to give you a heads up

File: read_buffer/src/row_group.rs [hunk 2]
```diff
@@ -195,16 +195,16 @@ impl RowGroup {
...
+    /// rows that satisfy the predicate, but does not materialise any
+    /// values. There are some optimisation opportunities here, but I don't
+    /// think they're at all worth it at the moment.
     ///
     /// They could be:
     ///  * for predicates with single expression just find a matching value in
     ///    the column;
     ///  * in some cases perhaps work row by row rather than column by column.
-    pub fn satisfies_predicate(&self, predicate: &Predicate) -> bool {
+    pub fn row_satisfies_predicate(&self, predicate: &Predicate) -> bool {
```
Comment (alamb) [inline read_buffer/src/row_group.rs:207]: FWIW I expected a function named `row_satisfies_predicate` to work on a single row

I suggest either leaving the name as `satisfies_predicate` or make it `row_group_satisfies_predicate` for additional readability

File: read_buffer/src/column.rs [hunk 3]
```diff
@@ -1088,10 +1136,30 @@ pub enum IntegerEncoding {
...
+        }
+    }
+
+    /// Determines if the column contains a non-null value at one of the
+    /// provided rows.
+    pub fn has_non_null_value(&self, row_ids: &[u32]) -> bool {
+        match self {
+            IntegerEncoding::I64I64N(enc) => enc.has_non_null_value(row_ids),
+            IntegerEncoding::U64U64N(enc) => enc.has_non_null_value(row_ids),
+            _ => !row_ids.is_empty(), // all rows will be non-null
```
Comment (alamb) [inline read_buffer/src/column.rs:1161]: As a guard against forgetting to update this case (and the ones above it) if additional Nullable integer encodings are added, it might make sense to explicitly enumerate all the possible encoding types -- that way if a new encoding is added the compiler will require adding a new branch explicitly

**Comment by e-dard** [comment]
@alamb 

> Do the results have 3 or 4 rows? (aka is the empty line meant to signify a row with "NULL" or that the result does not contain any row for counter?

Sorry for the terrible example (in hindsight!). Three rows returned. The empty line is just room for the `<--` comment. Full results would be:

```
names
------- 
env
region
time```

File: read_buffer/src/lib.rs (hunk 1 above)
Comment (e-dard) [reply read_buffer/src/lib.rs:512]: Seems fine to me.

File: read_buffer/src/column.rs (hunk 3 above)
Comment (e-dard) [reply read_buffer/src/column.rs:1161]: That's a good idea and actually the cause of a bug I fixed in another place in the PR. I shall do this for all of my enums I think from now on!

File: read_buffer/src/row_group.rs [hunk 4]
```diff
@@ -922,6 +922,51 @@ impl RowGroup {
...
         dst.aggregates.push(AggregateResults(aggregate_row)); // write the row
     }
+
+    /// Given the optional predicate, determines a set of rows contained in this
+    /// row group that satisfy it. Any column that contains a non-null value at
+    /// any of these row positions is then included in the results, which are
+    /// added to `dst`.
+    ///
+    /// As an optimisation, the contents of `dst` are checked before execution
+    /// and any columns already existing in the set are not interrogated.
```
Comment (carols10cents) [inline read_buffer/src/row_group.rs:932]: Errrr.... this seems like really surprising behavior to me? I see the test you have too, that affirms this behavior, that if you pass in a `dst` containing column names `["track", "time"]` and ask for the column names from a row group that contains only `["env", "time"]` columns, that you get back `["env", "track", "time"]`, but if you pass in an empty `dst` you'll get back `["env", "time"]`... (also what's the ordering? seems arbitrary? is that correct?)... I guess I can see that being useful for efficiently finding the UNION of column name sets for different row groups, and that if you want totally independent calls you should always pass in an empty `dst`..... I'm not sure of my point here, I guess my point is that this surprised me, but maybe it won't surprise others.

File: read_buffer/src/chunk.rs [hunk 5]
```diff
@@ -257,18 +257,27 @@ impl Chunk {
...
-        found_keys: &BTreeSet<ColumnName<'_>>,
-    ) -> BTreeSet<ColumnName<'_>> {
-        // Lookup table by name and dispatch execution if the table's time range
-        // overlaps the requested time range *and* there exists columns in the
-        // table's schema that are *not* already found.
-        todo!();
+        table_name: &str,
+        predicate: &Predicate,
+        dst: BTreeSet<String>,
+    ) -> BTreeSet<String> {
```
Comment (carols10cents) [inline read_buffer/src/chunk.rs:272]: Hm, why do `Chunk`'s and `Table`s `column_names` take ownership of `dst` and return it, while `RowGroup`s `column_names` takes a `&mut dst`?

File: read_buffer/src/lib.rs [hunk 6]
```diff
@@ -470,20 +474,42 @@ impl Database {
...
+            })?;
+
+        let chunk_data = partition.data.read().unwrap();
+        let mut filtered_chunks = vec![];
+        for id in chunk_ids {
+            filtered_chunks.push(
+                chunk_data
+                    .chunks
+                    .get(id)
+                    .ok_or_else(|| Error::ChunkNotFound { id: *id })?,
```
Comment (carols10cents) [inline read_buffer/src/lib.rs:497]: This will return an error for the first `chunk_id` not found... would it be better if this returned *all* the chunk ids that weren't found?

File: read_buffer/src/lib.rs [hunk 7]
```diff
@@ -470,20 +474,42 @@ impl Database {
...
+
+        let partition = partition_data
+            .partitions
+            .get(partition_key)
+            .ok_or_else(|| Error::PartitionNotFound {
+                key: partition_key.to_owned(),
+            })?;
+
+        let chunk_data = partition.data.read().unwrap();
+        let mut filtered_chunks = vec![];
```
Comment (carols10cents) [inline read_buffer/src/lib.rs:491]: I see you need to use a `for` loop rather than `map`/`collect` so that you can use `?`, but I think this could save some potential reallocations by at least doing:

```suggestion
        let mut filtered_chunks = Vec::with_capacity(chunk_ids.len());
```

File: read_buffer/src/lib.rs [hunk 8]
```diff
@@ -470,20 +474,42 @@ impl Database {
...
+        }
+
+        let names = filtered_chunks
+            .iter()
+            .fold(BTreeSet::new(), |dst, chunk| {
+                // the dst buffer is pushed into each chunk's `column_names`
+                // implementation ensuring that we short-circuit any tables where
+                // we have already determined column names.
+                chunk.column_names(table_name, &predicate, dst)
+            }) // have a BTreeSet here, convert to an iterator of Some(&str)
```
Comment (carols10cents) [inline read_buffer/src/lib.rs:508]: Actually, I wonder if this logic can be incorporated into the `for` loop above and avoid allocating the `filtered_chunks` vector entirely? I'd work up a proposal but I think it depends on your answer to whether we'd want to return *all* not found chunk ids or not...

**Review by carols10cents — COMMENTED** [review]
Some questions and suggestions!

**Comment by e-dard** [comment]
Thanks for the review @carols10cents! I will fix these issues but it may not be done until thu/fri.

File: read_buffer/src/row_group.rs (hunk 4 above)
Comment (e-dard) [reply read_buffer/src/row_group.rs:932]: @carols10cents perhaps this can be resolved with documentation? Yes that's the behaviour - we want to build a union of results so it is inefficient to waste time getting a column result if we already have got it previously.

File: read_buffer/src/chunk.rs (hunk 5 above)
Comment (e-dard) [reply read_buffer/src/chunk.rs:272]: which is preferable for you? I think I changed my mind and preferred the former (return owned content) but I forgot to update `RowGroup`.

File: read_buffer/src/lib.rs (hunk 6 above)
Comment (e-dard) [reply read_buffer/src/lib.rs:497]: I think returning the first one is fine. The caller should handle what to do (in fact the caller should be surprised if they see this error because they're supposed to know what chunks are in the read buffer).

File: read_buffer/src/lib.rs (hunk 8 above)
Comment (e-dard) [reply read_buffer/src/lib.rs:508]: I don't want all not found chunk ids at the moment.

File: read_buffer/src/column/dictionary/plain.rs [hunk 9]
```diff
@@ -664,13 +664,29 @@ impl Plain {
...
-        todo!()
+        if !self.contains_null() {
+            return true;
+        }
+
+        // If there are encoded values at any of the row IDs then there is at
+        // least one non-null value in the column.
+        row_ids
+            .iter()
+            .any(|id| self.encoded_data[*id as usize] != NULL_ID)
```
[the remark this answers is not in the document]
Comment (e-dard) [reply read_buffer/src/column/dictionary/plain.rs:689]: Yes I saw these inconsistencies during this branch development and made a note to fix them up. I decided that we should panic if a out-of-bound row ID is provided. I'll fix in a subsequent PR.

File: read_buffer/src/lib.rs (hunk 6 above)
Comment (alamb) [reply read_buffer/src/lib.rs:497]: Yeah, I agree -- it is a bug in a higher layer (in the `server` crate somewhere) if an invalid chunk_id is passed in
